File: Lua/FractionNative.lua
-- Useful utilities
local function round(value)
  return math.floor(value + 0.5)
end

local function inherit(klass,parent,metamethods)
  if klass.mt ~= nil and parent.mt ~= nil then
    for _,k in ipairs(metamethods) do
      if parent.mt[k] ~= nil and klass.mt[k] == nil then
        klass.mt[k] = parent.mt[k]
      end
    end
    klass.mt.__index = parent.mt
  end
end

local function isa(obj,tbl_or_type)
--if (type(obj) == tbl_or_type)
  if type(obj) == "table" then
    mt = getmetatable(obj)
    if mt then
      while true do
        if rawequal(mt,tbl_or_type.mt) then
          return true
        end
        mt_index=mt.__index
        if(mt_index == nil) or (rawequal(mt_index,mt)) then
          return false
        end
        mt=mt_index
      end
    end
  end
  return false
end

local function to_i(value)
  local sign = 1
  if value < 0 then
    sign = -1
    value = -value
  end
  return sign*math.floor(value)
end

-- Fraction class
Fraction = {  epsilon = 5e-6 , loops = 0 , __name = "Fraction" }
Fraction.__index = Fraction
Fraction.mt = { }
Fraction.mt.__index=Fraction.mt

function Fraction:new(...)
  local o = { numerator = 0, denominator = 1}
  setmetatable(o,Fraction.mt)
  o:set(...)
  return o
end

-- Creates a fraction of the specified type (normal Fraction or Mixed Mraction)
-- Actually, create a table and assign the specified metatable
local function fraction_of_same_type(self,...)
  local mt = getmetatable(self)
  o = Fraction:new(...)
  setmetatable(o,mt)
  return o
end

function Fraction.gcd(a,b)
  a=math.abs(a)
  b=math.abs(b)
  while(b ~= 0) do
    local t = b
    b = a % b
    a = t
  end
  return a
end

function Fraction.mt:reduce()
  if (self.denominator < 0) then
    self.denominator = -self.denominator
    self.numerator = -self.numerator
  end
  local gcd=Fraction.gcd(self.numerator,self.denominator)
  if (gcd ~= 1) then
    self.denominator = self.denominator/gcd
    self.numerator = self.numerator/gcd
  end
  return self
end

function Fraction.mt:set_mixed(w,n,d)
  sign = 1
  w=to_i(w)
  n=to_i(n)
  d=to_i(d)
  if(w < 0) then
    sign = sign*-1
    w=-w
  end
  if(n < 0) then
    sign = sign*-1
    n=-n
  end
  if(d < 0) then
    sign = sign*-1
    d=-d
  end
  self.numerator=sign*(w*d+n)
  self.denominator=d
  self:reduce()
end

function Fraction.mt:set_fraction(value)
  local is_fraction = false
  local w,n,d
  w,n,d = value:match("^%s*([-+]?%d+)%s+([-+]?%d+)/([-+]?%d+)%s*$")
  if w and n and d then
    self:set_mixed(tonumber(w),tonumber(n),tonumber(d))
    is_fraction = true
  else
    n,d = value:match("^%s*([-+]?%d+)/([-+]?%d+)%s*$")
    if n and d then
      self.numerator=tonumber(n)
      self.denominator=tonumber(d)
      self:reduce()
      is_fraction = true
    end
  end
  return is_fraction
end

function Fraction.mt:set_number(value)
  local hm2=0
  local hm1=1
  local h=0
  local km2=1
  local km1=0
  local k=0
  local v = value
  Fraction.loops = 0
  while(true) do
    local a = to_i(v)
    h=a*hm1 + hm2
    k=a*km1 + km2
    if(math.abs(value - h/k) < Fraction.epsilon) then
      break
    end
    v = 1.0/(v -a)
    hm2=hm1
    hm1=h
    km2=km1
    km1=k
    Fraction.loops = Fraction.loops + 1
  end
  if(k < 0) then
    k=-k
    h=-h
  end
  self.numerator=h
  self.denominator=k
end

function Fraction.mt:set_string(value)
  local ok = false
  local value_as_number=tonumber(value)
  if value_as_number then
    ok = true
    self:set_number(value_as_number)
  else
    ok = self:set_fraction(value)
  end
  return ok
end

function Fraction.mt:set(...)
  local args = {...}
  local nargs = #args
  if(nargs == 0) then
    self.numerator=0
    self.denominator=1
  elseif(nargs == 1) then
    if isa(args[1],Fraction) then
      self.numerator=args[1].numerator
      self.denominator=args[1].denominator
    elseif type(args[1]) == "table"  then
      self:set(table.unpack(args[1]))
    elseif(type(args[1])=="number") then
      self:set_number(args[1])
    elseif(type(args[1])=="string") then
      if not self:set_string(args[1]) then
        -- error
      end
    end
  elseif(nargs==2) then
    self.numerator=to_i(args[1])
    self.denominator=to_i(args[2])
    self:reduce()
  elseif(nargs==3) then
    self:set_mixed(args[1],args[2],args[3])
  else
    -- error
  end
  return self
end

function Fraction.mt:round(d)
  local f_new = fraction_of_same_type(self,self.numerator,self.denominator)
--  setmetatable(f_new,mt)
  d=math.abs(d)
  if(d < self.denominator) then
    f_new.numerator = round((f_new.numerator*d/f_new.denominator))
    f_new.denominator = d
    f_new:reduce()
  end
  return f_new
end

function Fraction.mt:abs()
  local fnew = fraction_of_same_type(self,self)
  fnew.numerator = math.abs(fnew.numerator)
  return fnew
end

function Fraction.mt:tonumber()
  return self.numerator/self.denominator
end

local function value_to_fraction(value)
  if isa(value,Fraction) then
    -- Already a fraction
    return value
  end
  if type(value) == "number" then
    return Fraction:new(value)
  end
  return nil
end

function Fraction.mt:__add(o)
  local result = nil
  if isa(self,Fraction) then
    local rhs = value_to_fraction(o)
    if rhs then
      result = fraction_of_same_type(self,self.numerator*rhs.denominator+rhs.numerator*self.denominator,self.denominator*rhs.denominator)
    end
  elseif (type(self) == "number") then
    result = self + o.numerator/o.denominator
  end
  return result
end

function Fraction.mt:__sub(o)
  local result = nil
  if isa(self,Fraction) then
    local rhs = value_to_fraction(o)
    if rhs then
      result = fraction_of_same_type(self,self.numerator*rhs.denominator-rhs.numerator*self.denominator,self.denominator*rhs.denominator)
    end
  elseif (type(self) == "number") then
    result = self - o.numerator/o.denominator
  end
  return result
end

function Fraction.mt:__mul(o)
  local result = nil
  if isa(self,Fraction) then
    local rhs = value_to_fraction(o)
    if rhs then
      result = fraction_of_same_type(self,self.numerator*rhs.numerator,self.denominator*rhs.denominator)
    end
  elseif (type(self) == "number") then
    result = self * o.numerator/o.denominator
  end
  return result
end

function Fraction.mt:__div(o)
  local result = nil
  if isa(self,Fraction) then
    local rhs = value_to_fraction(o)
    if rhs then
      result = fraction_of_same_type(self,self.numerator*rhs.denominator,rhs.numerator*self.denominator)
    end
  elseif (type(self) == "number") then
    result = self / o.numerator/o.denominator
  end
  return result
end

function Fraction.mt:__idiv(o)
  local result = nil
  if isa(self,Fraction) then
    local rhs = value_to_fraction(o)
    if rhs then
      result = self / rhs
      result.numerator=math.floor(result.numerator/result.denominator)
      result.denominator=1
    end
  elseif (type(self) == "number") then
    result = self // (o.numerator/o.denominator)
  end
  return result
end

local function fraction_to_number(value)
  if isa(value,Fraction) then
    return value.numerator/value.denominator
  end
  if type(value) =="number" then
    return value
  end
  return nil
end

function Fraction.mt:__pow(o)
  local result = nil
  local b = fraction_to_number(self)
  local e = fraction_to_number(o)
  if b < 0 then
    if to_i(e) ~= e then
      -- error
    end
  end
  result = b ^ math.abs(e)
  if e < 0 then
    result = 1.0/result
  end
  if isa(self,Fraction) then
    local temp = fraction_of_same_type(self,self)
    result = temp:set(result)
  end
  return result
end

function Fraction.mt:__unm(o)
  result = fraction_of_same_type(self,self)
  result.numerator = -result.numerator
  return result
end

function Fraction.mt:__len()
  return self.numerator/self.denominator
end

local function fraction_cmp_fraction(lhs,rhs)
  local l = lhs.numerator*rhs.denominator
  local r = rhs.numerator*lhs.denominator
  if l < r then
    return -1
  end
  if l > r then
    return 1
  end
  return 0
end

function Fraction.mt:__eq(o)
  print(self.numerator,self.denominator,o)
  return fraction_cmp_fraction(value_to_fraction(self),value_to_fraction(o)) == 0
end

function Fraction.mt:__lt(o)
  return fraction_cmp_fraction(value_to_fraction(self),value_to_fraction(o)) < 0
end

function Fraction.mt:__le(o)
  return fraction_cmp_fraction(value_to_fraction(self),value_to_fraction(o)) <= 0
end

function Fraction.mt:__tostring()
  s=string.format("%d",self.numerator)
  if(self.denominator ~= 1) then
    s=s..string.format("/%d",self.denominator)
  end
  return s
end

MixedFraction = {  __name = "MixedFraction" , __index = Fraction }
MixedFraction.mt = {  }
MixedFraction.mt.__index = Fraction.mt

function MixedFraction.mt:__tostring()
  s=nil
  if(math.abs(self.numerator) > self.denominator and self.denominator ~= 1) then
    w = to_i(self.numerator/self.denominator)
    n = math.abs(self.numerator) - math.abs(w)*self.denominator
    s = string.format("%d %d/%d",w,n,self.denominator)
  else
    s=Fraction.mt.__tostring(self)
  end
  return s
end

local fraction_metamethods_to_inherit = {
      "__lt",
      "__div",
      "__mul",
      "__eq",
      "__tostring",
      "__len",
      "__le",
      "__unm",
      "__idiv",
      "__add",
      "__pow",
      "__sub"
}

inherit(MixedFraction,Fraction,fraction_metamethods_to_inherit)
setmetatable(MixedFraction,Fraction)

function MixedFraction.new(self,...)
  o = Fraction:new(...)
  setmetatable(o,MixedFraction.mt)
  return o
end

File: Python/fraction_native.py
import re

class Fraction:
  'Fraction class'
  epsilon = 5e-6
  loops = 0
  __re_fraction = re.compile(r"^\s*([-+]?\d+?\s+)?([-+]?\d+)\/([-+]?\d+)\s*$")

  @staticmethod
  def isnumber(n):
    return isinstance(n,int) or isinstance(n,float)

  @staticmethod
  def isint(n):
    return (int(n) == n)

  @staticmethod
  def GCD(aa,bb):
    a=abs(aa)
    b=abs(bb)
    while b != 0:
      t = b
      b = a % b
      a = t

    return a

  @staticmethod
  def reduce(f):
    if f.__denominator < 0:
      f.__denominator = -f.__denominator
      f.__numerator = -f.__numerator
    gcd = Fraction.GCD(f.__numerator,f.__denominator)
    f.__denominator /= gcd
    f.__numerator /= gcd

  @staticmethod
  def from_mixed(f,w,n,d):
    sign = 1
    if w < 0:
      w=-w
      sign*=-1
    if n < 0:
      n=-n
      sign*=-1
    if d < 0:
      sign*=-1
      d = -d
    f.__numerator = sign*(w*d+n)
    f.__denominator = d
    Fraction.reduce(f)

  @staticmethod
  def from_number(f,value):
    hm2=0
    hm1=1
    h=0
    km2 = 1
    km1 = 0
    k=0
    Fraction.loops = 0
    v = value
    while True:
      a = int(v)
      h = a*hm1 + hm2
      k = a*km1 + km2
      if(abs(value - float(h)/float(k)) < Fraction.epsilon):
        break
      v = 1.0/(v - float(a))
      hm2 = hm1
      hm1 = h
      km2 = km1
      km1 = k
      Fraction.loops += 1
    if k < 0:
      k=-k
      h=-h
    f.__numerator = h
    f.__denominator = k

  @staticmethod
  def from_fraction_string(f,s):
    m =Fraction.__re_fraction.match(s)
    valid = False
    if isinstance(m,re.Match):
      valid = True
      if m.group(1) == None:
        f.__numerator = int(m.group(2))
        f.__denominator = int(m.group(3))
      else:
        Fraction.from_mixed(f,int(m.group(1)),int(m.group(2)),int(m.group(3)));
    return valid

  @staticmethod
  def from_number_string(f,s):
    valid=True
    try:
      Fraction.from_number(f,float(s))
    except ValueError:
      valid=False
    return valid

  @staticmethod
  def from_string(f,s):
    valid = Fraction.from_fraction_string(f,s)
    if not valid:
      valid = Fraction.from_number_string(f,s)
    return valid

  @staticmethod
  def value_to_fraction(value):
    if isinstance(value,Fraction):
      return value
    if Fraction.isnumber(value):
      return Fraction(value)
    return NotImplemented

  @staticmethod
  def fraction_to_value(value):
    if isinatance(value,Fraction):
      return float(value.__numerator)/float(value.__denominator)
    if Fraction.isnumber(value):
      return value
    return NotImplemented

  @classmethod
  def Loops(self):
    return Fraction.loops

  def __init__(self,*args):
    Fraction.epsilon=5e-6
    self.set(*args)

  @property
  def numerator(self):
    return self.__numerator

  @property
  def denominator(self):
    return self.__denominator

  def set(self,*args):
    nargs=len(args)
    self.__numerator = 0
    self.__denominator = 1
    if nargs == 0:
      self.__numerator = 0
      self.__denominator = 1
    elif nargs == 1:
      if isinstance(args[0],Fraction):
        self.__numerator = args[0].__numerator
        self.__denominator = args[0].__denominator
      elif isinstance(args[0],int):
        self.__numerator = args[0]
        self.__denominator = 1
      elif isinstance(args[0],float):
        Fraction.from_number(self,args[0])
      elif isinstance(args[0],str):
        if not Fraction.from_string(self,args[0]):
          print("Exception of invalid string")
      else:
        print("Exception")

    elif nargs == 2:
      if Fraction.isint(args[0]) and Fraction.isint(args[1]):
        self.__numerator = int(args[0])
        self.__denominator = int(args[1])
        Fraction.reduce(self)

    elif nargs == 3:
      if Fraction.isint(args[0]) and Fraction.isint(args[1]) and Fraction.isint(args[2]):
        Fraction.from_mixed(self,int(args[0]),int(args[1]),int(args[2]))

    else:
      print("Error")

  def __round__(self,denom=100):
    v = 0
    if denom < self.__denominator:
      v = type(self)(int(round(float(self.__numerator)*float(denom)/float(self.__denominator))),int(denom))
    else:
      v = type(self)(self)
    return v

  def __abs__(self):
    fnew = type(self)(self)
    fnew.__numerator = abs(fnew.__numerator)
    return fnew

  def __hash__(self):
    return hash((self.__numerator,self.__denominator))

  def __float__(self):
    return float(self.__numerator)/float(self.__denominator)

  def __int__(self):
    return int(Fraction.__float__(self))

  def __trunc__(self):
    return Fraction.__int__(self)

  def __floor__(self):
    return floor(Fraction.__float__(self))

  def __ceil__(self):
    return ceil(Fraction.__float__(self))

  def __str__(self):
    s= "%d" % (self.__numerator)
    if self.__denominator !=1:
      s+="/%d" % (self.__denominator)
    return s

  def __format__(self,format_spec):
    return format(str(self), format_spec)

  def __eq__(self,other):
    lhs = Fraction.value_to_fraction(self)
    rhs = Fraction.value_to_fraction(other)
    return lhs.__numerator*rhs.__denominator == lhs.__denominator*rhs.__numerator

  def __ne__(self,other):
    lhs = Fraction.value_to_fraction(self)
    rhs = Fraction.value_to_fraction(other)
    return lhs.__numerator*rhs.__denominator != lhs.__denominator*rhs.__numerator

  def __lt__(self,other):
    lhs = Fraction.value_to_fraction(self)
    rhs = Fraction.value_to_fraction(other)
    return lhs.__numerator*rhs.__denominator < lhs.__denominator*rhs.__numerator

  def __le__(self,other):
    lhs = Fraction.value_to_fraction(self)
    rhs = Fraction.value_to_fraction(other)
    return lhs.__numerator*rhs.__denominator <= lhs.__denominator*rhs.__numerator

  def __gt__(self,other):
    lhs = Fraction.value_to_fraction(self)
    rhs = Fraction.value_to_fraction(other)
    return lhs.__numerator*rhs.__denominator > lhs.__denominator*rhs.__numerator

  def __ge__(self,other):
    lhs = Fraction.value_to_fraction(self)
    rhs = Fraction.value_to_fraction(other)
    return lhs.__numerator*rhs.__denominator >= lhs.__denominator*rhs.__numerator

  def __add__(self,other):
    if isinstance(other,Fraction) or Fraction.isnumber(other):
      rhs = Fraction.value_to_fraction(other)
      return type(self)(int(self.__numerator*rhs.__denominator + rhs.__numerator*self.__denominator),int(self.__denominator*rhs.__denominator))
    return NotImplemented

  def __radd__(self,other):
    if Fraction.isnumber(other):
      return other + float(self.__numerator)/float(self.denominator)
    return NotImplemented

  def __sub__(self,other):
    if isinstance(other,Fraction) or Fraction.isnumber(other):
      rhs = Fraction.value_to_fraction(other)
      return type(self)(int(self.__numerator*rhs.__denominator - rhs.__numerator*self.__denominator),int(self.__denominator*rhs.__denominator))
    return NotImplemented

  def __rsub__(self,other):
    if Fraction.isnumber(other):
      return other - float(self.__numerator)/float(self.denominator)
    return NotImplemented

  def __mul__(self,other):
    if isinstance(other,Fraction) or Fraction.isnumber(other):
      rhs = Fraction.value_to_fraction(other)
      return type(self)(int(self.__numerator*rhs.__numerator),int(self.__denominator*rhs.__denominator))
    return NotImplemented

  def __rmul__(self,other):
    if Fraction.isnumber(other):
      return other * float(self.__numerator)/float(self.denominator)
    return NotImplemented

  def __truediv__(self,other):
    if isinstance(other,Fraction) or Fraction.isnumber(other):
      rhs = Fraction.value_to_fraction(other)
      return type(self)(int(self.__numerator*rhs.__denominator),int(rhs.__numerator*self.__denominator))
    return NotImplemented

  def __rtruediv__(self,other):
    if Fraction.isnumber(other):
      return other*float(self.denominator) / float(self.__numerator)
    return NotImplemented

  def __pow__(self,other):
    if not Fraction.isnumber(other) and not isinstance(other,Fraction):
      return NotImplemented
    b = float(self)
    e = float(other)
    if b < 0 and not Fraction.isint(e):
      return NotImplemented
    r=pow(b,abs(e))
    if e<0:
      r=1.0/r
    return type(self)(r)

  def __rpow__(self,other):
    if not Fraction.isnumber(other):
      return NotImplemented
    b = float(other)
    e = float(self)
    if b < 0 and not Fraction.isint(e):
      return NotImplemented
    r=pow(b,abs(e))
    if e<0:
      r=1.0/r
    return r

class MixedFraction(Fraction):

  def __str__(self):
    if self.denominator == 1 or abs(self.numerator) < self.denominator:
      return Fraction.__str__(self)
    w = self.numerator / self.denominator
    n = abs(self.numerator) % self.denominator
    return "%d %d/%d" % (w,n,self.denominator)
